### PR TITLE
feat(layers): GasEstimationLayer

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Ref: #323 

## Solution

Implement a `layer` that populates the `gas_price`, `gas`, `max_fee_per_gas` and `max_priority_fee_per_gas`.

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/provider/src/layers/gas.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            to: address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into(),
+            ..Default::default()
+        };
+
+        let tx = provider.send_transaction(tx).await.unwrap();
+
+        let tx = tx.get_receipt().await.unwrap();
+
+        let set_gas_price = U128::from(0x77359400);
+        let set_gas_limit = U256::from(0x5208);
```
Comment (gakonst) [inline crates/provider/src/layers/gas.rs:121]: We should make the tx type have the `u128` and `u64` types where possible i think instead of `U128`?

File: crates/provider/src/layers/gas.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        mut tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<'_, N, T>> {
+        if tx.gas_price().is_none() {
+            let gas_price = self.get_gas_price().await?;
+            tx.set_gas_price(gas_price);
+        }
+        if tx.gas_limit().is_none() {
+            let gas_estimate = self.get_gas_estimate(&tx).await?;
+            tx.set_gas_limit(gas_estimate);
+        }
```
Comment (gakonst) [inline crates/provider/src/layers/gas.rs:84]: this should be done with `match` and `join`'ing the 2 futures if both are present, to make it faster.

i also think we should do this for `nonce`, and perhaps this should be a general layer that fills both all fields

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/provider/src/layers/gas.rs (hunk 1 above)
Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:135]: Could be done but I think that should be a separate PR. Bunch of places where we're using U128 and U64.

File: crates/provider/src/layers/gas.rs (hunk 2 above)
Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:84]: > i also think we should do this for `nonce`, and perhaps this should be a general layer that fills both all fields
So make generalized tx builder layer and do nonce management and gas estimation in one?

Comment (gakonst) [reply crates/provider/src/layers/gas.rs:84]: Yes I think so cc @prestwich @DaniPopes @mattsse for their view

File: crates/provider/src/layers/gas.rs (hunk 1 above)
Comment (gakonst) [reply crates/provider/src/layers/gas.rs:135]: yeah separate OK

File: crates/provider/src/layers/gas.rs (hunk 2 above)
Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs]: Using `join!` now.

**Comment by yash-atreya** [comment]
Decided to make this a more generic tx builder layer. Ref: https://github.com/alloy-rs/alloy/pull/326#discussion_r1526791175

Renaming it to `FillTxLayer`.
This layer will manage nonces and estimates the gas.

**Comment by yash-atreya** [comment]
> Decided to make this a more generic tx builder layer. Ref: [#326 (comment)](https://github.com/alloy-rs/alloy/pull/326#discussion_r1526791175)
> 
> Renaming it to `FillTxLayer`. This layer will manage nonces and estimates the gas.

Undo above.

File: crates/provider/src/layers/gas.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    async fn send_transaction(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<'_, N, T>> {
+        let (gas_price, gas_limit, eip1559_fees) = match (
+            tx.gas_price().is_none(),
+            tx.gas_limit().is_none(),
+            (tx.max_fee_per_gas().is_none() && tx.max_priority_fee_per_gas().is_none()),
+        ) {
+            (true, true, true) => {
```
Comment (onbjerg) [inline crates/provider/src/layers/gas.rs:82]: pls add comments to make it clearer what tx types we are filling

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/provider/src/layers/gas.rs (hunk 3 above)
Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:82]: @onbjerg by tx types do you mean `EIP1559`, or `legacy` or do you mean specifying the fields being populated?

Comment (prestwich) [reply crates/provider/src/layers/gas.rs:82]: for this, we can assume that there are 2 intended ways:
- 1559 style (assumed default)
- legacy style (IF and ONLY IF the user has already set the `gas_price` key).

the builder should have a type-level guarantee that `gas_price` and the 1559 gas fields are mutually exclusive, i.e. `(true,true,true)` should never be permitted at all.

as a result we should be able to simplify the 8 match arms to an `if let Some(price) = tx.gas_price` or similar

Comment (prestwich) [reply crates/provider/src/layers/gas.rs:82]: in the future, we might want to allow a network-associated `GasStrategy` to handle this 🤔

Comment (onbjerg) [reply crates/provider/src/layers/gas.rs:82]: I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set a gas price, in which case, we probably shouldn't override it?

Comment (onbjerg) [reply crates/provider/src/layers/gas.rs:82]: > @onbjerg by tx types do you mean EIP1559, or legacy or do you mean specifying the fields being populated?

i mean, please add comments on the arms to explain what we are matching for, it is quite hard to parse what `(true, true, false)` etc means

Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:82]: > I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set the `gas_price` field, in which case, we probably shouldn't override it?

Agree with @onbjerg's, point here. 

How about?
```rust
if tx.gas_price().is_none() {
   // Assume its a EIP1559 tx
   // Populate the following fields: gas_limit, max_fee_per_gas and max_priority_fee_per_gas 
} else {
   // Assume its a legacy tx
   // Populate only the gas_limit field if unset.
 }
```

Comment (prestwich) [reply crates/provider/src/layers/gas.rs:82]: > > I don't quite understand this part though, if this is supposed to fill `gas_price` or `max_fee_per_gas`/`max_prio_fee_per_gas`, would they not always be unset? i.e. it would never be a legacy tx unless you set the `gas_price` field, in which case, we probably shouldn't override it?
> 
> Agree with @onbjerg's, point here.
> 
> How about?
> 
> ```rust
> if tx.gas_price().is_none() {
>    // Assume its a EIP1559 tx
>    // Populate the following fields: gas_limit, max_fee_per_gas and max_priority_fee_per_gas 
> } else {
>    // Assume its a legacy tx
>    // Populate only the gas_limit field if unset.
>  }
> ```

Yes, this is the correct strategy. Recommend decomposing into `fill_legacy` and `fill_1559` functions

Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs]: Pushed changes. 

In this update, if the `gas_price` is not set (i.e assumed to be 1559 tx) but the network doesn't support eip1559 txs then it falls back to being process as a legacy tx.

File: crates/provider/src/layers/gas.rs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    ) -> Result<N::TransactionRequest, TransportError> {
+        // Check if the user has set the gas_limit, max_fee_per_gas and max_priority_fee_per_gas
+        // fields.
+        let gas_limit_unset = tx.gas_limit().is_none();
+        // eip1559_fees_unset is true if any one of the two max_fee_per_gas or
+        // max_priority_fee_per_gas is unset.
+        let eip1559_fees_unset =
+            tx.max_fee_per_gas().is_none() || tx.max_priority_fee_per_gas().is_none();
+
+        match (gas_limit_unset, eip1559_fees_unset) {
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:108]: remove the combinatoric match please

follow the pattern in `populate_gas` and `populate_gas_eip1559` in `provider.rs`

File: crates/provider/src/layers/gas.rs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                Ok(tx)
+            }
+        }
+    }
+
+    /// Populates the gas_price and only populates the gas_limit field if unset.
+    /// This method always assumes that the gas_price is unset.
+    async fn handle_legacy_tx(
+        &self,
+        mut tx: N::TransactionRequest,
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:199]: `&mut`

File: crates/provider/src/layers/gas.rs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+    async fn get_eip1559_fees_estimate(&self) -> TransportResult<(U256, U256)> {
+        self.inner.estimate_eip1559_fees(None).await
+    }
+
+    /// Populates the gas_limit, max_fee_per_gas and max_priority_fee_per_gas fields if unset.
+    /// Requires the chain_id to be set in the transaction request to be processed as a EIP-1559 tx.
+    /// If the network does not support EIP-1559, it will process it as a legacy tx.
+    async fn handle_eip1559_tx(
+        &self,
+        mut tx: N::TransactionRequest,
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:98]: &mut

File: crates/provider/src/layers/gas.rs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    async fn get_eip1559_fees_estimate(&self) -> TransportResult<(U256, U256)> {
+        self.inner.estimate_eip1559_fees(None).await
+    }
+
+    /// Populates the gas_limit, max_fee_per_gas and max_priority_fee_per_gas fields if unset.
+    /// Requires the chain_id to be set in the transaction request to be processed as a EIP-1559 tx.
+    /// If the network does not support EIP-1559, it will process it as a legacy tx.
+    async fn handle_eip1559_tx<'a>(
+        &'a self,
+        tx: &'a mut N::TransactionRequest,
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:98]: these must have different lifetimes

File: crates/provider/src/layers/gas.rs [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+                Ok(tx)
+            }
+        }
+    }
+
+    /// Populates the gas_price and only populates the gas_limit field if unset.
+    /// This method always assumes that the gas_price is unset.
+    async fn handle_legacy_tx<'a>(
+        &'a self,
+        tx: &'a mut N::TransactionRequest,
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:199]: different lifetimes

File: crates/provider/src/layers/gas.rs [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        }
+    }
+
+    /// Populates the gas_price and only populates the gas_limit field if unset.
+    /// This method always assumes that the gas_price is unset.
+    async fn handle_legacy_tx<'a, 'b>(
+        &'a self,
+        tx: &'b mut N::TransactionRequest,
+    ) -> Result<&'b mut N::TransactionRequest, TransportError> {
+        let gas_price = self.get_gas_price();
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:135]: here's a nice little pattern for future usage:

```
        let gas_limit_fut = if let Some(gas_limit) = tx.gas_limit() {
            futures::future::Either::Left(async move { gas_limit })
        } else {
            futures::future::Either::Right(async { self.estimate_gas(&*tx, None).await })
        };
```

Comment (prestwich) [reply crates/provider/src/layers/gas.rs:135]: with FuturesExt trait

```
let gas_limit_fut = tx.gas_limit() {
            async move { gas_limit }.left_future()
        } else {
            async { self.estimate_gas(&*tx, None).await }.right_future()
        };
```

Comment (prestwich) [reply crates/provider/src/layers/gas.rs:135]: that way you can run the `join!` without all the clunky if block

Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:147]: This was a really cool thing I learned! Thanks.

File: crates/provider/src/layers/gas.rs [hunk 10]
```diff
@@ -144,30 +144,20 @@ where
...
-            match gas_price {
-                Ok(gas_price) => {
-                    tx.set_gas_price(gas_price);
-                    Ok(tx)
-                }
-                Err(err) => Err(err),
+            async { self.get_gas_estimate(tx).await }.right_future()
+        };
+
+        match futures::try_join!(gas_price_fut, gas_limit_fut) {
```
Comment (prestwich) [inline crates/provider/src/layers/gas.rs:154]: this `match` block is equivalent to a `.map` right?

Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:154]: Oh yes!

File: crates/provider/src/layers/gas.rs [hunk 11]
```diff
@@ -144,30 +144,20 @@ where
         &'a self,
         tx: &'b mut N::TransactionRequest,
     ) -> Result<&'b mut N::TransactionRequest, TransportError> {
```
[the remark this answers is not in the document]
Comment (yash-atreya) [reply crates/provider/src/layers/gas.rs:146]: The arg is `&mut`, and returning a `N::TransactionRequest` would require cloning.

**Review by prestwich — APPROVED** [review]

File: crates/provider/src/layers/gas.rs [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        match futures::try_join!(gas_estimate_fut, eip1559_fees_fut) {
+            Ok((gas_limit, eip1559_fees)) => {
+                tx.set_gas_limit(gas_limit);
+                tx.set_max_fee_per_gas(eip1559_fees.max_fee_per_gas);
+                tx.set_max_priority_fee_per_gas(eip1559_fees.max_priority_fee_per_gas);
+                Ok(())
+            }
+            Err(err) => {
+                if err.is_transport_error()
+                    && err.to_string() == *"EIP-1559 not activated".to_string()
```
Comment (gakonst) [inline crates/provider/src/layers/gas.rs:127]: I'd love this to a helper method on the error re 1559 but we can do as followup

**Review by gakonst — APPROVED** [review]

File: crates/provider/src/layers/gas.rs (hunk 12 above)
Comment (prestwich) [reply crates/provider/src/layers/gas.rs:127]: actually this is a bit of a hole in the API. 1559 not activated should not be a transport error. "endpoint doesn't exist" falls in a hole between transport error and rpc error 🤔
